### PR TITLE
Deleting users by the District Admin

### LINKED ISSUE #1309: Add an option to delete users
District admin should be able to delete others users

### PR DESCRIPTION
**Pull request by Sandra3657**
Resolves #1309 

The district admins can now delete users of the user types which the district admins can create. The district admins cannot delete users, that are above them (State Admin, State Lab Admin etc).

delete

### CODE REVIEW

File: src/Components/Users/ManageUsers.tsx [hunk 1]
```diff
@@ -26,24 +36,37 @@ export default function ManageUsers(props: any) {
   const [totalCount, setTotalCount] = useState(0);
   const [currentPage, setCurrentPage] = useState(1);
   const [offset, setOffset] = useState(0);
+  const [reload, setReload] = useState<boolean>(false);
 
   const state: any = useSelector((state) => state);
   const { currentUser } = state;
   const isSuperuser = currentUser.data.is_superuser;
   const userType = currentUser.data.user_type;
+  // console.log(isSuperuser, userType);
```
Comment (smartclash) [inline src/Components/Users/ManageUsers.tsx]: Let's remove this

File: src/Components/Users/ManageUsers.tsx [hunk 2]
```diff
@@ -156,31 +193,81 @@ export default function ManageUsers(props: any) {
...
+    let res = await dispatch(deleteUser(username));
+    if (res.status >= 200) {
+      Notification.Success({
+        msg: "User deleted successfully",
+      });
+    }
+    console.log(userData);
+
+    setUserData({ show: false, username: "", name: "" });
+    // setReload(!reload);
```
Comment (smartclash) [inline src/Components/Users/ManageUsers.tsx]: and this too

File: src/Components/Users/ManageUsers.tsx [hunk 3]
```diff
@@ -156,31 +193,81 @@ export default function ManageUsers(props: any) {
...
+    window.location.reload();
+  };
+
+  const handleDelete = (user: any) => {
+    console.log(user);
```
Comment (smartclash) [inline src/Components/Users/ManageUsers.tsx]: and this

File: src/Components/Users/ManageUsers.tsx [hunk 4]
```diff
@@ -156,31 +193,81 @@ export default function ManageUsers(props: any) {
...
   const showLinkFacility = (username: string) => {
-    return <a onClick={() => showLinkFacilityModal(username)} className={`align-baseline font-bold text-sm ${!isFacilityLoading ? "text-blue-500 hover:text-blue-800" : "text-gray-500"}`} href="#" >Link new facility</a>;
+    return (
+      <a
+        onClick={() => showLinkFacilityModal(username)}
+        className={`align-baseline font-bold text-sm ${
+          !isFacilityLoading
+            ? "text-blue-500 hover:text-blue-800"
+            : "text-gray-500"
+        }`}
```
Comment (smartclash) [inline src/Components/Users/ManageUsers.tsx]: Can we use `classnames` package here?

File: src/Components/Users/ManageUsers.tsx [hunk 5]
```diff
@@ -156,31 +193,81 @@ export default function ManageUsers(props: any) {
...
-          <DeleteForeverIcon />
-        </IconButton>
-      </div>)
-      ))}
-      {showLinkFacility(username)}
-    </>);
+    return (
+      <>
+        {facilities.map((facility) => (
+          <div className="flex items-center mb-2">
```
Comment (smartclash) [inline src/Components/Users/ManageUsers.tsx]: Please add a `key` here.

File: src/Components/Users/ManageUsers.tsx [hunk 6]
```diff
@@ -194,52 +281,99 @@ export default function ManageUsers(props: any) {
   let userList: any[] = [];
   if (users && users.length) {
     userList = users.map((user: any, idx: number) => {
+      console.log(user);
```
Comment (smartclash) [inline src/Components/Users/ManageUsers.tsx]: Let's remove this

File: src/Components/Users/ManageUsers.tsx [hunk 7]
```diff
@@ -156,31 +193,81 @@ export default function ManageUsers(props: any) {
...
+
+  const handleSubmit = async () => {
+    let username = userData.username;
+    let res = await dispatch(deleteUser(username));
+    if (res.status >= 200) {
+      Notification.Success({
+        msg: "User deleted successfully",
+      });
+    }
+    console.log(userData);
```
Comment (smartclash) [inline src/Components/Users/ManageUsers.tsx]: and this

**Review by smartclash — CHANGES_REQUESTED** [review]

File: src/Components/Users/ManageUsers.tsx (hunk 4 above)
Comment (Sandra3657) [reply src/Components/Users/ManageUsers.tsx]: Yes. Have added it.

**Review by smartclash — APPROVED** [review]

**Comment by amruthavijay88** [comment]
Tested, unable to delete someone of staff user type. Warning comes - which says 'Not found' and then the user seems to exist after deletion.

**Comment by shaileshaanand** [comment]
I can delete Staff users from my local backend but not from staging

**Comment by Sandra3657** [comment]
Can you say which user is that? I have tested it out with a usertype of Staff and I am able to delete the user. @amruthavijay88

**Comment by amruthavijay88** [comment]
Tested again, unable to delete users with 'staff' user access in staging again.

**Comment by amruthavijay88** [comment]
@Sandra3657 @shaileshaanand

**Comment by nihal467** [comment]
@Sandra3657 

step 1: Create a user id for example "test-arun"
step 2: delete the staff user-id "test-arun"
step 3: create again the same user-id "test-arun"

You will have a notification that 'user-id already exists' but in the front end, you will not be able to see the user id.

**Comment by bodhish** [comment]
@nihal467 we do a soft delete. You can't re-use a user id that's already used.

**Comment by bodhish** [comment]
image


When I tried to delete a district admin the API crashed. Do handle these edge cases.

**Comment by bodhish** [comment]
@vigneshhari Any idea why I am getting a 404 for the delete endpoint 🤔

**Comment by bodhish** [comment]
Also Update 

```
  deleteUser: {
    path: "/api/v1/users",
    method: "DELETE",
  },
```
with

```
  deleteUser: {
    path: "/api/v1/users/",
    method: "DELETE",
  },

```

api.txs

**Comment by vigneshhari** [comment]
@bodhish 404 is thrown the user does not have access to delete that user , dist admin trying to delete other districts dist admin/staff or a state admin or cases like that.

**Comment by bodhish** [comment]
That's nice. Please hide the delete button for those users

**Comment by Sandra3657** [comment]
For a district admin, the dropdown for adding new users contains `DistrictAdmin` and `DistrictReadOnlyAdmin` as an option too. So if a district admin cannot delete another district admin, he shouldn't be able to able to create one too, right? @bodhish

**Comment by bodhish** [comment]
Delete should be down by a user with one level more access.

**Comment by aparnacoronasafe** [comment]
Screenshot 2021-06-09 at 12 45 12 PM

When trying to delete staff, "Something went wrong" notification comes up

**Comment by Sandra3657** [comment]
It was working in the deploy preview, soon after the last commit. It is still working in my local server, although not in the deploy preview now. @aparnacoronasafe @bodhish

**Comment by nihal467** [comment]
error1

Test failed, not able to delete a user with 'staff' role

**Comment by aparnacoronasafe** [comment]
> It was working in the deploy preview, soon after the last commit. It is still working in my local server, although not in the deploy preview now. @aparnacoronasafe @bodhish

@bodhish Please look at this

**Comment by bodhish** [comment]
@Sandra3657 get help from cef in slack

**Comment by arpancodes** [comment]
Implemented the following:
- [x] District admin can delete the users below it in the same district
- [x] State Admin can delete users from the same state
- [x] User below the District Admin won't be able to see the delete button
- [x] In rare case, if the user tries to delete a user that they don't have the permission for, `permission denied` alert will be shown.

Please test @amruthavijay88 @aparnacoronasafe

**Comment by nihal467** [comment]
errorror


test failed, with a district admin access, when I am trying to delete a staff user, a pop-up alert 'permission denied'

**Comment by arpancodes** [comment]
This is only happening in deploy preview! I tried deleting same staff in both local server and deploy preview. Check this out:
In local server:
delete_staff_localhost

in deploy preview:
delete_staff_deploy_preview

**Comment by nihal467** [comment]
test failed, using a state admin access, I can delete another user with state admin  profile

**Comment by arpancodes** [comment]
@nihal467 if you were able to delete that means it's possible from the server else you'd see a permission denied, should I disable the delete button just on the frontend?

**Comment by nihal467** [comment]
@vigneshhari  @bodhish check this issue

**Comment by vigneshhari** [comment]
You can delete other stateadmins provided they are in the same state.

**Comment by aparnacoronasafe** [comment]
cannot find the option in preview

**Comment by arpancodes** [comment]
@aparnacoronasafe You can only see delete button for the one you have access to delete for example, a district admins can delete the ones with lower access but in the same district.
image

**Comment by aparnacoronasafe** [comment]
@arpancodes my bad!! yes tested in preview

**Comment by bodhish** [comment]
@arpancodes can you resolve the merge conflicts 
Good to merge after that

**Comment by amruthavijay88** [comment]
Tested in Netlify. Logged in as a district admin, can't seem to find delete button? 

image

**Comment by aparnacoronasafe** [comment]
@amruthavijay88 

Check Arpans comment above...:-/

**Comment by amruthavijay88** [comment]
@aparnacoronasafe - My bad, didn't notice it was another district's staff.

**Comment by arpancodes** [comment]
@bodhish yaa will do that

**Pull request merged**